luarules/gadgets/game_team_power_watcher.lua

```diff
@@ -62,233 +29,462 @@ if gadgetHandler:IsSyncedCode() then
+    UnitsWithPower = {},
+
+    
+    --teams
+    ScavengerTeam = nil,
+    RaptorTeam = nil,
+    NeutralTeam = nil,
+    AITeams = {},
+    HumanTeams = {}
+}
+
+local spGetGameSeconds = Spring.GetGameSeconds
+
+local teamList = Spring.GetTeamList()
 
+local testPowerTable = {}
+local testPowerNumber = 0
 
 --decipher human vs ai vs neutral vs defense mode ai's 
-    for _, teamID in ipairs(teamList) do
-		local teamLuaAI = Spring.GetTeamLuaAI(teamID)
-		if (teamLuaAI and string.find(teamLuaAI, "ScavengersAI")) then
-			scavTeam = teamID
-        elseif (teamLuaAI and string.find(teamLuaAI, "RaptorsAI")) then
-            raptorTeam = teamID
-        elseif select (4, Spring.GetTeamInfo(teamID, false)) then
-            AIteams[teamID] = true
-        elseif teamID == tonumber(teamList[#teamList]) then
-            neutralTeamNumber = teamID
-        else
-			humanTeams[teamID] = true
-		end
-	end
+for _, teamID in ipairs(teamList) do
+    local teamLuaAI = Spring.GetTeamLuaAI(teamID)
+    if (teamLuaAI and string.find(teamLuaAI, "ScavengersAI")) then
+        GG.PowerLib.ScavengerTeam = teamID
+    elseif (teamLuaAI and string.find(teamLuaAI, "RaptorsAI")) then
+        GG.PowerLib.RaptorTeam = teamID
+    elseif select (4, Spring.GetTeamInfo(teamID, false)) then
+        GG.PowerLib.AITeams[teamID] = true
+    elseif teamID == tonumber(teamList[#teamList]) then
+        GG.PowerLib.NeutralTeam = teamID
+    else
+        GG.PowerLib.HumanTeams[teamID] = true
+    end
+end
 
 --assign all teams power and peak power of 0 to prevent nil errors
-    for _, teamNumber in ipairs(teamList) do
-        teamPowers[teamNumber] = 0
-        peakTeamPowers[teamNumber] = 0
-    end
+for _, teamNumber in ipairs(teamList) do
+    GG.PowerLib.TeamPowers[teamNumber] = 0
+    GG.PowerLib.PeakTeamPowers[teamNumber] = 0
+end
 
 
 
-	function gadget:UnitFinished(unitID, unitDefID, unitTeam)
-		if UnitDefs[unitDefID].power then
-            unitsWithPower[unitID] = {unitID = unitID, power = UnitDefs[unitDefID].power, team = unitTeam}
-            teamPowers[unitTeam] = (teamPowers[unitTeam] + UnitDefs[unitDefID].power) or UnitDefs[unitDefID].power
-            highestTeamPower = TPW_HighestTeamPower()
-            averageTeamPower = TPW_AverageTeamPower()
-            averageHumanTeamPower = TPW_AverageHumanTeamPower(unitTeam)
-            averageAlliedTeamPower = TPW_AverageAlliedTeamPower(unitTeam)
-            averageTechGuesstimate = TPW_AverageTechGuesstimate()
-            averageAlliedTechGuesstimate = TPW_AverageAlliedTechGuesstimate(unitTeam)
-            averageHumanTechGuesstimate = TPW_AverageHumanTechGuesstimate()
+function gadget:UnitFinished(unitID, unitDefID, unitTeam)
+    GG.PowerLib.UnitsWithPower[unitID] = {unitID = unitID, power = UnitDefs[unitDefID].power, team = unitTeam}
+    GG.PowerLib.TeamPowers[unitTeam] = (GG.PowerLib.TeamPowers[unitTeam] + UnitDefs[unitDefID].power) or UnitDefs[unitDefID].power
+    
+    --debug
```

Comment:
Please remove debug and test code before merging

Reply:
Every function has been tested. I want those lines there because if I add a function this is the most optimal way I've found to text it quickly. Future generations will appreciate it IMO